Rework `Simulator` Reset API & back-end ops

vastsoun opened the pull request:
## Description
This PR mostly brings modifications to the internal methods of the `Simulator` to improve the resetting operations.

## Before your PR is "Ready for review"
- [x] Necessary tests have been added and new examples are tested (see `newton/tests/test_examples.py`)
- [x] Documentation is up-to-date
- [x] Code passes formatting and linting checks with `pre-commit run -a`

Guirec-Maloisel commented on newton/_src/solvers/kamino/simulation/simulator.py:
```diff
@@ -625,16 +718,23 @@ def sync_host(self):
     ###
 
-    def _run_reset_callback(self):
+    def _run_pre_reset_callback(self):
```
Wondering if this wrapper is needed, rather than just having the if directly in calling code
(applies to the others below as well)

vastsoun replied:
I wanted to avoid copy-paste as much as possible, plus, in the beginning some of these ran other operations after the callback. Let's just leave it for now and refine later.

vastsoun commented on newton/_src/solvers/kamino/simulation/resets.py:
```diff
@@ -261,7 +262,102 @@ def _reset_joints_of_select_worlds(
+        model: Input model container holding the time-invariant data of the system.
+        state: Input state container specifying the target state to be reset to.
+        mask: Array of per-world flags indicating which worlds should be reset.
+        data: Output solver data to be configured for the target state.
```
Fixed

vastsoun commented on newton/_src/solvers/kamino/simulation/simulator.py:
```diff
@@ -764,8 +859,30 @@ def _reset_all_worlds_to_initial_state(self):
+
+    def _reset_select_worlds_to_initial_state(self, world_mask: wp.array | None = None, reset_constraints: bool = True):
+        """
+        Resets the simulation to the initial state defined in the model.
```
Fixed

vastsoun commented:
> Looks good to me overall, just a few minor comments

@Guirec-Maloisel Thanks a lot for catching all these details I missed. I will push a final commit with fixes addressing all of your comments and proceed with merging.
